### ISSUE TITLE
twisted.protocols.ftp.FTP.ftp_STOR() does not handle anonymous requests gracefully

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11106]
|<img alt="grimmtooth's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| grimmtooth reported|
|-|-|
|Trac ID|trac#1575|
|Type|defect|
|Created|2006-03-15 05:36:05Z|

Around line 997 in ftp_STOR(), the protocol tries to fire off ```self.shell.openForWriting()``` without first checking to see if that method exists. In the case of an anonymous user, that would be ```FTPAnonymousShell``` which certainly doesn't have such a method. As a result, the remote user sees an "internal server error" as an exception gets raised on the server.

I'd be inclined to suggest returning ```NEED_ACCT_FOR_STOR``` except I can envision other cases where the shell might not support that method without being an actual anon login. Maybe simply ```PERMISSION_DENIED```?

Resulting traceback:

```
     File "E:\Python\Lib\site-packages\twisted\internet\tcp.py", line 349, in doRead
       return self.protocol.dataReceived(data)
     File "E:\Python\Lib\site-packages\twisted\protocols\policies.py", line 72, in dataReceived
       self.wrappedProtocol.dataReceived(data)
     File "E:\Python\Lib\site-packages\twisted\protocols\basic.py", line 232, in dataReceived
       why = self.lineReceived(line)
     File "E:\Python\lib\site-packages\twisted\protocols\ftp.py", line 642, in lineReceived
       d = defer.maybeDeferred(self.processCommand, cmd, *args)
   --- <exception caught here> ---
     File "E:\Python\Lib\site-packages\twisted\internet\defer.py", line 107, in maybeDeferred
       result = f(*args, **kw)
     File "E:\Python\lib\site-packages\twisted\protocols\ftp.py", line 673, in processCommand
       return method(*params)
     File "E:\Python\lib\site-packages\twisted\protocols\ftp.py", line 997, in ftp_STOR
       d = self.shell.openForWriting(newsegs)
   exceptions.AttributeError: 'FTPAnonymousShell' object has no attribute 'openForWriting'
```

Attachments:

* [ftp.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/863/8639c2e8c44d79a5afaf095aa46f9c9231829b72/0de399c06f87caeab252614aa73c0f7c0ff3108d.diff) (1090 bytes) - added by grimmtooth on 2006-03-16 03:51:39Z - Possible fix. Please review to ensure it meets your criterea.

<details><summary>Searchable metadata</summary>

```
trac-id__1575 1575
type__defect defect
reporter__grimmtooth grimmtooth
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ftp ftp
keywords__ 
time__1142400965000000 1142400965000000
changetime__1177834772000000 1177834772000000
version__ 
owner__ 
cc__spiv cc__grimmtooth cc__therve cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="grimmtooth's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|grimmtooth commented|
|-|-|

Corrected spelling error in title

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv set owner to spiv<br>spiv set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve<br>@therve set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set owner to spiv|
|-|-|

I used a different fix in ftp-anon-stor-1575, by implementing the method in FTPAnonymousShell. Please review.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_8"></a>|@mwhudson commented|
|-|-|

At the very least, please fix the english in the docstring: the attempt results *in* an exception, not *to* an exception.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve commented|
|-|-|

Thanks, corrected.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @therve|
|-|-|

I adjusted the docstring for the new openForWriting method.

The interface says that openForWriting should return a Deferred, and ftp_STOR doesn't seem to account for openForWriting raising a synchronous exception (but lineReceived accounts for processCommand raising a synchronous exception, and ftp_STOR is called from processCommand).  I think the behavior is being supplied somewhat by accident.  Probably openForWriting should return defer.fail(...) and ebOpened in ftp_STOR should account for PermissionDeniedError differently than it does now (eg, by returning PERMISSION_DENIED (or PermissionDeniedError.errorCode) instead of FILE_NOT_FOUND).

There's a defer.setDebugging(True) call left in FTPServerTestCase.setUp that should be removed.

The new test case FTPServerTestCaseAdvancedClient is about as good as the existing test coverage, but the existing test coverage isn't so hot.  If, in addition, you could add a test which instantiates FTPAnonymousShell directly, calls openForWriting on it, and asserts that it returns a deferred that fails in the appropriate way, that would be excellent.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve removed owner|
|-|-|

OK, that's probably better.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set owner to @therve|
|-|-|

In ebOpened:

```
            if isinstance(err.value, FTPCmdError):
```

should be:

```
            if err.check(FTPCmdError):
```

Otherwise, looks great.  Please merge after fixing the above.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve set status to `closed`|
|-|-|

(In [20098]) Merge ftp-anon-stor-1575

Author: therve
Reviewer: exarkun
Fixes [#11106](https://github.com/twisted/twisted/issues/11106)

Implement openForWriting in FTPAnonymousShell: this allows the FTP server to
give a significant error when trying STOR in anonymous.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
